lib/vrl/compiler/src/compiler.rs

```diff
@@ -395,33 +337,35 @@ impl<'a> Compiler<'a> {
     fn compile_predicate(
         &mut self,
         node: Node<ast::Predicate>,
-        external: &mut ExternalEnv,
+        state: &mut TypeState,
     ) -> Option<predicate::Result> {
         use ast::Predicate::{Many, One};
 
         let (span, predicate) = node.take();
 
         let exprs = match predicate {
-            One(node) => vec![self.compile_expr(*node, external)?],
-            Many(nodes) => self.compile_exprs(nodes, external)?,
+            One(node) => vec![self.compile_expr(*node, state)?],
+            Many(nodes) => self.compile_exprs(nodes, state)?,
         };
 
         Some(Predicate::new(
             Node::new(span, exprs),
-            (&self.local, external),
+            state,
             self.fallible_expression_error.as_deref(),
         ))
     }
 
     #[cfg(feature = "expr-op")]
-    fn compile_op(&mut self, node: Node<ast::Op>, external: &mut ExternalEnv) -> Option<Op> {
+    fn compile_op(&mut self, node: Node<ast::Op>, state: &mut TypeState) -> Option<Op> {
         use parser::ast::Opcode;
 
+        let original_state = state.clone();
```

Comment:
I'll add comments with this info, but I'll also explain the overall thought process here.
This changes here required each expression to be able to fully calculate the type definition changes that it makes itself. This code lives in the `Expression::type_info` function. The compiler _also_ needs to keep track of the current state of the type system as it's compiling, mostly for fallibility checking. Rather than duplicating this code in both the compiler and expressions, the compiler now calls the `type_info` function when needed.

All of the `Compiler::compile_*` functions have a mutable `state` argument which is the state of the program right before the current expression would execute. This state should be modified as if the compiled expression had executed. In places where the state isn't naturally updated to the correct value, updating the state is delegated to the `type_info` function of the compiled expression instead.

The best example for this is `compile_if_statement`. There are a few places where the state is reset, and a single branch of the if statement is compiled. At the end of this function, the state needs to reflect the entire expression, so the original state is passed into `if_statement.type_info(..)` and _that_ state is used as the new compiler state.

Reply:
I would like to eventually get the compiler to a place where manually updating the state isn't necessary and it can happen automatically from the `type_info` function. I'll look into that in a future PR though.